### PR TITLE
Code for AAAI 2021 paper. "SCNet: Training Inference Sample Consistency for Instance Segmentation"

### PR DESCRIPTION
**Pull request by thangvubk**
# Short description: 
- We propose [SCNet](https://arxiv.org/abs/2012.10150) to address the **Sample Inconsistency** problem in existing **Cascade method** for instance segmentation. 
- SCNet is achieves **2.3 points mask AP improvement** on strong baseline Cascade Mask R-CNN while **running faster and requires less memory** (with 16G RAM GPU, we are able to train SCNet-ResNext-101 with batch-size of 2 per GPU).
- Experiments on the paper are conducted on mmdet1, here we provide implementation on mmdet2 with slightly better results.

### CODE REVIEW

**Comment by hellock** [comment]
Thanks for the contribution!

**Comment by xvjiarui** [comment]
Hi @thangvubk 
Unit test failed. Could you please fix it?

File: mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py [hunk 1]
```diff
@@ -170,7 +174,10 @@ def forward(self, x):
 
         cls_score = self.fc_cls(x_cls) if self.with_cls else None
         bbox_pred = self.fc_reg(x_reg) if self.with_reg else None
-        return cls_score, bbox_pred
+        if return_feat:
+            return cls_score, bbox_pred, feature
+        else:
+            return cls_score, bbox_pred
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py:180]: I suggest refactoring this function into two parts, `forward_shared`, `forward_cls_reg`. So we don't need `return_feat` argument.

File: mmdet/models/roi_heads/mask_heads/fcn_mask_head.py [hunk 2]
```diff
@@ -53,22 +55,34 @@ def __init__(self,
...
-                self.in_channels if i == 0 else self.conv_out_channels)
-            padding = (self.conv_kernel_size - 1) // 2
-            self.convs.append(
-                ConvModule(
-                    in_channels,
-                    self.conv_out_channels,
-                    self.conv_kernel_size,
-                    padding=padding,
-                    conv_cfg=conv_cfg,
-                    norm_cfg=norm_cfg))
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/mask_heads/fcn_mask_head.py:71]: I suggest adding a new mask head inherited `FCNMaskHead` instead.

File: mmdet/models/roi_heads/mask_heads/global_context_head.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+from mmdet.models.utils import ResLayer, SimplifiedBasicBlock
+
+
+@HEADS.register_module()
+class GlobalContextHead(nn.Module):
+    """Global context head used in SCNet.
+
+    https://arxiv.org/abs/2012.10150
+    """
+
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/mask_heads/global_context_head.py:15]: Missing docstring.

File: mmdet/models/roi_heads/mask_heads/feature_relay_head.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+from mmdet.models.builder import HEADS
+
+
+@HEADS.register_module()
+class FeatureRelayHead(nn.Module):
+    """Feature Relay Head used in SCNet.
+
+    https://arxiv.org/abs/2012.10150
+    """
+
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/mask_heads/feature_relay_head.py:14]: Missing docstring.

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+
+@HEADS.register_module()
+class SCNetRoIHead(CascadeRoIHead):
+    """RoIHead for SCNet.
+
+    https://arxiv.org/abs/2012.10150
+    """
+
+    def __init__(self,
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/scnet_roi_head.py:16]: Missing docstring.

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+        if self.with_feat_relay:
+            self.feat_relay_head.init_weights()
+
+    @property
+    def with_semantic(self):
+        """bool: whether the head has semantic head"""
+        if hasattr(self, 'semantic_head') and self.semantic_head is not None:
+            return True
+        else:
+            return False
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/scnet_roi_head.py:73]: Simplify into one line.

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+    @property
+    def with_feat_relay(self):
+        """bool: whether the head has feature relay head"""
+        if (hasattr(self, 'feat_relay_head')
+                and self.feat_relay_head is not None):
+            return True
+        else:
+            return False
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/scnet_roi_head.py:82]: Simplify into one line.

File: mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py (hunk 1 above)
Comment (thangvubk) [reply mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py]: I agree that using two functions will be more clear. However, ``return_shared_feat`` at the common ``forward()`` should be required.

**Comment by hellock** [comment]
Kindly ping @xvjiarui again.

File: configs/scnet/scnet_r50_fpn_20e_coco.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
+_base_ = [
+    '../_base_/datasets/coco_instance.py',
+    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
+]
```
Comment (xvjiarui) [inline configs/scnet/scnet_r50_fpn_20e_coco.py:4]: Is it possible to inherit configs from HTC of Cascade RCNN, to make config shorter?

Comment (thangvubk) [reply configs/scnet/scnet_r50_fpn_20e_coco.py:4]: Thanks. This is a nice point.

Comment (xvjiarui) [reply configs/scnet/scnet_r50_fpn_20e_coco.py:4]: Btw, although inherit SCRoIHead from HTCRoIHead won't save much code. I still suggest doing so since they share similar components. Inheritance may be beneficial for future refactoring.

File: mmdet/models/utils/res_layer.py [hunk 9]
```diff
@@ -100,3 +100,86 @@ def __init__(self,
                     norm_cfg=norm_cfg,
                     **kwargs))
         super(ResLayer, self).__init__(*layers)
+
+
+class SimplifiedBasicBlock(nn.Module):
```
Comment (xvjiarui) [inline mmdet/models/utils/res_layer.py:105]: Unit test for this module is missing.

File: mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py [hunk 10]
```diff
@@ -132,8 +132,8 @@ def init_weights(self):
                     nn.init.xavier_uniform_(m.weight)
                     nn.init.constant_(m.bias, 0)
 
-    def forward(self, x):
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/bbox_heads/convfc_bbox_head.py:135]: After discussing with @hellock and @ZwwWayne , we think we may keep `convfc_bbox_head.py` unchanged. 
Then, we could add a new box head for this function.

File: mmdet/models/roi_heads/mask_heads/fused_semantic_head.py [hunk 11]
```diff
@@ -4,6 +4,7 @@
 from mmcv.runner import auto_fp16, force_fp32
 
 from mmdet.models.builder import HEADS
+from mmdet.models.utils import ResLayer, SimplifiedBasicBlock
 
 
 @HEADS.register_module()
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/mask_heads/fused_semantic_head.py:10]: Similarly, we think we may keep `fused_semantic_head.py` unchanged.
Then, we could add a new mask head for the desired function.

File: mmdet/models/utils/res_layer.py [hunk 12]
```diff
@@ -100,3 +100,86 @@ def __init__(self,
...
+    """Simplified version of original basic residual block:
```
Comment (xvjiarui) [inline mmdet/models/utils/res_layer.py:106]: We may add SCNet reference in the docstring.

**Comment by thangvubk** [comment]
@xvjiarui thanks for review my PR. I have updated the PR with your comments. After review is finish, we will benchmark the models again with different backbones and update the Readme.

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+    """RoIHead for SCNet https://arxiv.org/abs/2012.10150.
+
+    Args:
+        num_stages (int): number of cascade stages.
+        stage_loss_weights (list): loss weight of cascade stages.
+        semantic_roi_extractor (dict): config to init semantic roi extractor.
+        semantic_head (dict): config to init semantic head.
+        feat_relay_head (dict): config to init feature_relay_head.
+        glbctx_head (dict): config to init global context head.
+    """
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/scnet_roi_head.py:20]: I think there is one missing [comment](https://github.com/open-mmlab/mmdetection/pull/4356#discussion_r554665621).

**Comment by xvjiarui** [comment]
Except for one comment above, it looks good to me.

File: mmdet/models/roi_heads/scnet_roi_head.py (hunk 13 above)
Comment (thangvubk) [reply mmdet/models/roi_heads/scnet_roi_head.py:20]: I think i have resolved the that comment. I only inherit the config HTC including he backbone, rpn, neck, train/test configs, data pipeline except for RoIHead. The SCNetRoIHead use very differrent design and ``forward()`` logics from HTCRoIHead. Besides, SCNetRoIHead does not have ``interleave`` and ``mask_info_flow`` as in HTC.

Comment (xvjiarui) [reply mmdet/models/roi_heads/scnet_roi_head.py:20]: I see. Thanks for the explain.

**Review by xvjiarui — APPROVED** [review]

**Comment by thangvubk** [comment]
Thanks for your review. 

TODO: update pretrained models:
- [ ] R50 1x
- [ ] R50 20e
- [ ] R101 20e
- [ ] X101 20e

**Comment by thangvubk** [comment]
@xvjiarui, I have finish my TOTO list and update readme. Feel free to have a look. Best.

**Comment by xvjiarui** [comment]
LGTM. @ZwwWayne may have a look to see if there is any additional comments.

File: configs/scnet/README.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
+# SCNet
+
```
Comment (ZwwWayne) [inline configs/scnet/README.md:2]: For algorithms, now the readme requires to add a tag [ALGORITHM] for calculating algorithm statistics. See example [here](https://github.com/open-mmlab/mmdetection/tree/master/configs/tridentnet).

File: mmdet/models/roi_heads/mask_heads/feature_relay_head.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
+import torch.nn as nn
+from mmcv.cnn import kaiming_init
+from mmcv.runner import auto_fp16
+
+from mmdet.models.builder import HEADS
+
+
+@HEADS.register_module()
+class FeatureRelayHead(nn.Module):
+    """Feature Relay Head used in SCNet https://arxiv.org/abs/2012.10150.
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/mask_heads/feature_relay_head.py:10]: May use 
```python
`SCNet <https://arxiv.org/abs/2012.10150>`_.
```
To make the link can be clicked.

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 16]
```diff
@@ -0,0 +1,585 @@
...
+
+from mmdet.core import (bbox2result, bbox2roi, bbox_mapping, merge_aug_bboxes,
+                        merge_aug_masks, multiclass_nms)
+from ..builder import HEADS, build_head, build_roi_extractor
+from .cascade_roi_head import CascadeRoIHead
+
+
+@HEADS.register_module()
+class SCNetRoIHead(CascadeRoIHead):
+    """RoIHead for SCNet https://arxiv.org/abs/2012.10150.
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/scnet_roi_head.py:12]: See other comments to update the link

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 17]
```diff
@@ -0,0 +1,585 @@
...
+        bbox_targets = bbox_head.get_targets(sampling_results, gt_bboxes,
+                                             gt_labels, rcnn_train_cfg)
+        loss_bbox = bbox_head.loss(bbox_results['cls_score'],
+                                   bbox_results['bbox_pred'], rois,
+                                   *bbox_targets)
+
+        bbox_results.update(
+            loss_bbox=loss_bbox,
+            rois=rois,
+            bbox_targets=bbox_targets,
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/scnet_roi_head.py:196]: This comma is unnecessary.

**Comment by ZwwWayne** [comment]
Hi @thangvubk,
Thanks for the kind contribution. We just update our final review and plan to merge this PR this week. Could you resolve the conflicts and final comments on this PR?

**Comment by thangvubk** [comment]
Thanks for your review. Lint will be fixed after #4539 is merged.
